ENH create a vendor version of the library

glemaitre opened the pull request:
This PR create a single-file version of the library that third-party can vendor easily without depending on the package:

### TODO

- [x] add tests to check the import works
- [x] add documentation the top of the file
- [x] rework the README and thus the website documentation
- [x] add a note in the contributing guide mentioning the duplication of the code

adrinjalali commented:
I don't like this. This means we have to maintain two versions of the same code.

I don't think this is a priority, but if we want to really provide this, I rather have a script which generates a single file and puts it somewhere automatically (in a gist or something), instead of doing it manually.

adrinjalali commented:
These are going to be out of sync REALLY fast, so having a tool which gives users a vendored version is much nicer. Also, if users want to vendor, they can vendor the `src/sklearn_compat` folder instead of a single file.

adrinjalali commented:
As library devs, you and I, I think are happy to depend on the actual package. I don't think maintaining a massive large file is a good experience. Shall we let the vendor thing become a thing when library maintainers start requesting it? As for now, I think `skops`, `fairlearn`, and `imbalanced-learn` would be okay with depending on this.

adrinjalali commented:
Hmm, in that case, let's do one source of truth in the `sklearn_compat` file, and call it `skcompat` since it's not the library name anymore, and remove the tests for import only files then?

You might also want to have a separate PR for the linter? I guess you've enabled black or isort in your IDE or env?

glemaitre commented:
> You might also want to have a separate PR for the linter? I guess you've enabled black or isort in your IDE or env?

I forgot to add the reformatting and sorting with ruff. But right now, those changes for reformatting are minimal because we anyway touch the import.

glemaitre commented:
> Hmm, in that case, let's do one source of truth in the sklearn_compat file, and call it skcompat since it's not the library name anymore, and remove the tests for import only files then?

So I added a `_sklearn_compat.py` file. I removed the duplication of the tests as well.

glemaitre commented:
And for the record, @adrinjalali I allow you to recall me "I told you so" when things will go sideways in a couple of versions downstream :)

adrinjalali commented on src/sklearn_compat/utils/_version.py:
delete the file then?

glemaitre replied:
Yep it was just generating `sklearn_version` that is in the vendor file. I just remove it then.

adrinjalali reviewed (approved):
LGTM, feel free to merge whenever you're done
